### ISSUE TITLE
Preserve terminal cwd when reviving process

### ISSUE THREAD

**Issue by craxal** [microsoft/vscode#200777]
Type: <b>Feature Request</b>

VS Code restores my terminal sessions on restarting, but it does not preserve the current working directories. I have to `cd` back into my cwd every time I restart Code. I would very much like the cwd to be preserved between sessions.

VS Code version: Code 1.85.1 (0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2, 2023-12-13T09:49:37.021Z)
OS version: Windows_NT x64 10.0.22621
Modes:

**Comment by meganrogge** [comment]
this works as long as you've typed in the terminal.

**Comment by meganrogge** [comment]
https://github.com/microsoft/vscode/assets/29464607/9aa25385-a70e-4f89-8649-3cb9238858fc

**Comment by craxal** [comment]
@meganrogge Not true. I'm not sure I see what your movie was meant to demonstrate, as I don't know what the cwd was supposed to be in the last session.

By default, the terminal opens to your home directory or the root of the current workspace. However, I can change the cwd at any time. It's this cwd that is not preserved between sessions:

![CwdNotPreserved](https://github.com/microsoft/vscode/assets/3473026/dc9f73b9-53c1-4e6c-96e3-41041648f092)

**Comment by meganrogge** [comment]
It's preserved on reload of the window, which is what my gif demonstrates. 

What do you have `terminal.integrated.persistentSessionReviveProcess` set to?

**Comment by craxal** [comment]
@meganrogge `onExit`

**Comment by craxal** [comment]
I'm not sure if that's the right value, but assuming it is, please reopen this issue.

**Comment by meganrogge** [comment]
as indicated above, use `onExitAndWindowClose` instead, which will have the effect you desire

**Comment by craxal** [comment]
@meganrogge It doesn't work for me:

![image](https://github.com/microsoft/vscode/assets/3473026/56f2bff5-e774-4fbd-bf96-98fde355c3da)

**Comment by meganrogge** [comment]
A gif would be helpful to see what you might be doing differently

**Comment by craxal** [comment]
@meganrogge The GIF I posted earlier demonstrates exactly what I'm doing. The only difference is I set `terminal.integrated.persistentSessionReviveProcess` to `onExitAndWindowClose` as you suggested.

**Comment by meganrogge** [comment]
I cannot reproduce this

**Comment by craxal** [comment]
@meganrogge The problem occurs both on Mac and Windows. I've been using pwsh on both platforms.

**Comment by Tyriar** [comment]
Related: https://github.com/microsoft/vscode/issues/234672

**Comment by meganrogge** [comment]
it's correctly set here
<img width="620" alt="Image" src="https://github.com/user-attachments/assets/4c7d1ee8-7ea0-43ce-873f-bd13f84178a9" />

and gets overwritten here
<img width="862" alt="Image" src="https://github.com/user-attachments/assets/672d8ec1-35fa-4999-993b-8307ad258419" />

**Comment by Tyriar** [comment]
@meganrogge is all we need to change that line to `??=`?

**Comment by meganrogge** [comment]
this is never entered for reconnected terminals

https://github.com/microsoft/vscode/blob/f940e6aaadea837bc32c6ae982f66c7a7c5ae73e/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L859

**Comment by meganrogge** [comment]
seems like we need to do something like this but for cwdDetection capability 

https://github.com/microsoft/vscode/blob/f940e6aaadea837bc32c6ae982f66c7a7c5ae73e/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L841

**Comment by meganrogge** [comment]
`shellIntegration.cwd` comes from the `CwdDetectionCapability` that we never reconnect to when a terminal restores. I think we need to serialize and deserialize that.

**Comment by meganrogge** [comment]
That gets created in `shellIntegrationAddon` when https://github.com/microsoft/vscode/blob/f940e6aaadea837bc32c6ae982f66c7a7c5ae73e/src/vs/platform/terminal/common/xterm/shellIntegrationAddon.ts#L492 this is called

**Comment by meganrogge** [comment]
edit: I see simpler fix, just make `updateCwd` public vs private in shell integration addon

**Comment by meganrogge** [comment]
i've got it close to working here, but `commandDetection` is not yet defined here so it doesn't get set there

<img width="1105" alt="Image" src="https://github.com/user-attachments/assets/e22ae92e-a183-49d0-acea-319210d810b8" />

**Comment by meganrogge** [comment]
seeing `shellIntegration` is udf for reconnected terminals 

![Image](https://github.com/user-attachments/assets/8f869f43-1f63-43f6-8d00-9124105465e7)

**Comment by craxal** [comment]
@rwoll This still reproduces for me.

**Comment by rwoll** [comment]
@craxal - here's the steps I'm performing:

1. Open VS Code and set `"terminal.integrated.persistentSessionReviveProcess": "onExitAndWindowClose"` at the User Preferences level.
2. On the command line, outside of VS Code create a new directory/file:
    ```
    mkdir -p repro/foo
    touch repro/foo/test.txt
    ```
2. Open the `repro` directory in VS Code.
3. Open the Terminal in VS Code.
4. Type `cd foo`
5. Type `pwd`
6. Close VS Code and re-launch.
7. Type `pwd` in the restored Terminal.

Both 6 + 7 print `…/repro/foo`.

Here's a video of me doing the above step:

https://github.com/user-attachments/assets/7764fc0d-de7e-42fc-8d2c-ea01a82d3806

**Comment by rwoll** [comment]
I worked with @Tyriar to understood the root issue. I've added the November milestone and will take the fix on during the iteration that starts next week as part of my onboarding.

Thanks for your patience!

**Comment by sean-mcmanus** [comment]
@craxal FYI, I was able to workaround this non-restoration of the cwd on close/open via using an extension (created by Copilot) to automatically open a new terminal with the desired cwd (and checking for a pre-existing one to avoid duplicates, such as in the reload window scenario where the terminal cwd can be preserved).
